client/network/src/discovery.rs

```diff
@@ -271,45 +342,61 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 		addr: &Multiaddr,
 		error: &dyn std::error::Error
 	) {
-		NetworkBehaviour::inject_addr_reach_failure(&mut self.kademlia, peer_id, addr, error)
+		for k in self.kademlias.values_mut() {
+			NetworkBehaviour::inject_addr_reach_failure(k, peer_id, addr, error)
+		}
 	}
 
 	fn inject_node_event(
 		&mut self,
 		peer_id: PeerId,
-		event: <Self::ProtocolsHandler as ProtocolsHandler>::OutEvent,
+		(pid, event): <Self::ProtocolsHandler as ProtocolsHandler>::OutEvent,
 	) {
-		NetworkBehaviour::inject_node_event(&mut self.kademlia, peer_id, event)
+		if let Some(kad) = self.kademlias.get_mut(&pid) {
+			kad.inject_node_event(peer_id, event)
+		}
```

Comment:
Let's not let bugs go potentially undetected!

```suggestion
		} else {
			error!(target: "sub-libp2p", "State inconsistency: Kademlia event with no associated behaviour");
		}
```

Same for all the other `for` loops below.